src/group_theory/index.lean

```diff
@@ -345,4 +345,54 @@ noncomputable def fintype_of_index_ne_zero (hH : H.index ≠ 0) : fintype (G ⧸
 lemma one_lt_index_of_ne_top [finite (G ⧸ H)] (hH : H ≠ ⊤) : 1 < H.index :=
 nat.one_lt_iff_ne_zero_and_ne_one.mpr ⟨index_ne_zero_of_finite, mt index_eq_one.mp hH⟩
 
+section finite_index
+
+variables (H K)
+
+/-- Typeclass for finite index subgroups. -/
+class finite_index : Prop :=
+(finite_index : H.index ≠ 0)
+
+/-- Typeclass for finite index subgroups. -/
+class _root_.add_subgroup.finite_index {G : Type*} [add_group G] (H : add_subgroup G) : Prop :=
+(finite_index : H.index ≠ 0)
+
+attribute [to_additive] finite_index
+
+@[priority 100, to_additive] noncomputable instance fintype_quotient_of_finite_index
+  [finite_index H] : fintype (G ⧸ H) :=
+fintype_of_index_ne_zero finite_index.finite_index
```

Comment:
Well, `subgroup.fintype_of_index_ne_zero` isn't an instance since it has the hypothesis `H.index ≠ 0`.

Having this `fintype` instance seems to be useful for the refactor. For instance, the transfer homomorphism currently requires `fintype`, and there are a few proofs that currently require `fintype` (e.g., in `schreier.lean`). And actually if you make this not an instance, then you can't switch `index_eq_card` to `finite_index` since it wouldn't be able to infer `fintype`.

Reply:
I'm still uneasy with creating instances from arbitrarily chosen data. I think for the transfer homomorphism you just need a `let : _ := fintype_quotient_of_finite_index H in by exactI` in the definition of `monoid_hom.transfer`, and you won't need `fintype_quotient_of_finite_index` as an instance in theorems involving `monoid_hom.transfer`.

Reply:
Maybe do that in [subgroup.left_transversals.diff](https://leanprover-community.github.io/mathlib_docs/group_theory/transfer.html#subgroup.left_transversals.diff) instead, since it's already noncomputable.

Reply:
I'll give it a shot, but it will probably also involve refactoring a few proofs that currently use `fintype` rather than `finite`.